core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -249,7 +249,14 @@ private[spark] class ExecutorAllocationManager(
    * yarn-client mode when AM re-registers after a failure.
    */
   def reset(): Unit = synchronized {
-    initializing = true
+    /**
+     * When some tasks need to be scheduled and initial executor = 0, resetting the initializing
+     * field may cause it to not be set to false in yarn.
+     * SPARK-20079: https://issues.apache.org/jira/browse/SPARK-20079
+     */
+    if (maxNumExecutorsNeeded() == 0) {
+      initializing = true
```

Comment:
`updateAndSyncNumExecutorsTarget` is a weird method. It returns a value that is never used anywhere, the actual variables it sets internally are what matters...

But I still don't understand why, when the AM restarts, should `updateAndSyncNumExecutorsTarget` be a no-op except in this case. What is different about this case that makes it an exception? Shouldn't `updateAndSyncNumExecutorsTarget` be called instead from `reset()` or very soon after, so the code can update its internal state to match the current status of the app?

The thing I don't understand is why is it ever ok for `updateAndSyncNumExecutorsTarget` to just do nothing.